### PR TITLE
feat(runner): implement all U32 arithmetic operations in VM

### PR DESCRIPTION
**Pull request by enitrat**
## Summary
- Implements all U32 arithmetic operations for Cairo-M VM runner with RISC-V-compliant behavior
- Adds comprehensive tests for all operations including edge cases

## Implementation Details

This PR implements all the U32 store operations defined in the Linear epic CORE-985:

### FP-FP Operations
- `U32StoreAddFpFp`: Addition of two U32 values from memory
- `U32StoreSubFpFp`: Subtraction of two U32 values from memory
- `U32StoreMulFpFp`: Multiplication of two U32 values from memory
- `U32StoreDivFpFp`: Division of two U32 values from memory

### FP-IMM Operations
- `U32StoreSubFpImm`: Subtraction with immediate value
- `U32StoreMulFpImm`: Multiplication with immediate value
- `U32StoreDivFpImm`: Division with immediate value

Note: `U32StoreAddFpImm` was already implemented.

### Key Design Decisions

1. **RISC-V-like operations**: All _except division_ operations follow RISC-V unsigned 32-bit arithmetic behavior:
   - All operations use wrapping arithmetic (no overflow traps)
   - Division by zero **panics instead**.

2. **Memory Layout**: U32 values are stored as two 16-bit M31 limbs in consecutive memory cells:
   - Limb0 (16 first bits) at `fp + offset`
   - Limb1 (16 last bits) at `fp + offset + 1`

3. **Validation**: All operations validate that limbs are within 16-bit range to prevent malformed data

## Test Plan
- [x] Unit tests for all U32 operations
- [x] Tests for edge cases (overflow, underflow, division by zero)
- [x] All tests pass (`cargo test -p cairo-m-runner`)
- [x] Code formatted with `trunk fmt`
- [x] Build succeeds (`cargo build -p cairo-m-runner`)

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: crates/runner/src/vm/instructions/store.rs [hunk 1]
```diff
@@ -181,40 +181,214 @@ pub fn store_div_fp_imm(
...
+    state: State,
+    instruction: &Instruction,
 ) -> Result<State, InstructionExecutionError> {
-    todo!("u32_store_add_fp_fp not implemented")
+    let (src0_off, src1_off, dst_off) =
+        extract_as!(instruction, U32StoreAddFpFp, (src0_off, src1_off, dst_off));
+
+    // Read first 32-bit value from memory as two limbs
+    let src0_limb_0 = memory.get_data(state.fp + src0_off)?;
+    let src0_limb_1 = memory.get_data(state.fp + src0_off + M31::from(1))?;
```
Comment (zmalatrax) [inline crates/runner/src/vm/instructions/store.rs:198]: `M31::one()` ?

File: crates/runner/src/vm/instructions/store.rs [hunk 2]
```diff
@@ -181,40 +181,214 @@ pub fn store_div_fp_imm(
...
+
+    // Validate that source limbs are within 16-bit range
+    if src0_limb_0.0 > U32_LIMB_MASK || src0_limb_1.0 > U32_LIMB_MASK {
```
Comment (zmalatrax) [inline crates/runner/src/vm/instructions/store.rs:201]: I don't think it's required, so we can consider removing them (perf-wise)
The prover will perform range checks on the limbs and throw an error anyway in that case 🤔

Comment (enitrat) [reply crates/runner/src/vm/instructions/store.rs:201]: We dont want to generate valid traces for invalid proofs

File: crates/runner/src/memory/mod.rs [hunk 3]
```diff
@@ -343,6 +356,31 @@ impl Memory {
             .flat_map(u32::to_le_bytes)
             .collect()
     }
+
+    /// Read a 32-bit value (little-endian) stored as two 16-bit limbs at `addr`.
+    pub fn read_u32(&self, addr: M31) -> Result<u32, MemoryError> {
```
Comment (zmalatrax) [inline crates/runner/src/memory/mod.rs:361]: Current convention is `get_x` (`get_instruction`, `get_data`)

I'm fine with `read_x`, but then we should update the other `get_x` in a subsequent PR (I'm refactoring the memory representation so I can add it to the scope of the refactor)

File: crates/runner/src/vm/instructions/store.rs [hunk 4]
```diff
@@ -3,84 +3,177 @@
...
+// Automatically-generated STORE FP-IMM operations (scalar)
+// -------------------------------------------------------------------------------------------------
+
+impl_store_bin_op_fp_imm!(store_add_fp_imm, StoreAddFpImm, +);
+impl_store_bin_op_fp_imm!(store_sub_fp_imm, StoreSubFpImm, -);
+impl_store_bin_op_fp_imm!(store_mul_fp_imm, StoreMulFpImm, *);
+impl_store_bin_op_fp_imm!(store_div_fp_imm, StoreDivFpImm, /);
+
+// -------------------------------------------------------------------------------------------------
+// Singular / less-regular STORE instructions (scalar)
```
Comment (zmalatrax) [inline crates/runner/src/vm/instructions/store.rs:175]: empty

File: crates/runner/src/vm/instructions/store_tests.rs [hunk 5]
```diff
@@ -1,188 +1,347 @@
...
+                dst_off: M31(2),
+            },
+            store_mul_fp_imm,
+            &[0, src_val, expected_res.0],
+            1,
+        ).unwrap();
+    }
+
+    #[test]
+    fn test_store_div_fp_fp(src0_val: u32, src1_val: u32) {
```
Comment (zmalatrax) [inline crates/runner/src/vm/instructions/store_tests.rs:294]: add a test that div by 0 throws an error ?

File: crates/runner/src/vm/instructions/store_tests.rs [hunk 6]
```diff
@@ -226,125 +366,120 @@ fn test_store_div_fp_imm_by_zero() {
...
-
-    Ok(())
+// -----------------------------------------------------------------------------
+// U32 FP-IMM instruction tests
+// -----------------------------------------------------------------------------
+
+proptest! {
+    #[test]
+    fn test_u32_store_add_fp_imm(src_value: u32, imm_val_hi in 0..=0xFFFFu32, imm_val_lo in 0..=0xFFFFu32) {
+        let imm_val = (imm_val_hi << 16) | imm_val_lo;
```
Comment (zmalatrax) [inline crates/runner/src/vm/instructions/store_tests.rs:376]: declare a constant as well for 16 ?

**Review by zmalatrax — APPROVED** [review]
